### PR TITLE
Direct support for html-webpack-plugin courtesy of a nice hook they added

### PR DESCRIPTION
**Pull request by SPSpwetter**
Also updated docs. 

With this change, anyone using html-webpack-plugin will get integrity attributes added to their script tags. 

Also bumped revision number. 

This also addresses issue #3

### CODE REVIEW

**Comment by jscheid** [comment]
Hey @SPSpwetter, nice one, looks great! Thanks for this.

`npm run lint` fails, could you fix it? I think you will have to upgrade `eslint`, `babel-eslint` and `eslint-plugin-react` first. Apologies for the README not mentioning linting, I'll improve that.

It would be helpful if your comment on html-webpack-plugin version requirements could somehow be codified, best practise appears to be [this](https://docs.npmjs.com/files/package.json#optionaldependencies), although emitting a warning rather than silently ignoring the dependency is probably the way to go. Ideally using the webpack error reporting facilities. What do you think?

And could you take a look at adding a simple test case? It would be nice if the code was at least executed by the test cases, and ideally a regexp test to ensure the hashes made it into index.html.

Thanks again, even though we're not using this plugin I can see this being useful to a lot of people.

**Comment by jscheid** [comment]
Also, could you remove the version bump? It should be a bump to `0.4.0` as per semver, but I'd prefer to do that separately.

**Comment by SPSpwetter** [comment]
Done:

* eslint issues
* version bump. Back to 0.3.2

Not Done:

* Test case. Not sure how to do this. 
* optionalDependencies change: I don't agree that optionalDependencies is the way to do this, as I think npm tries to install it, which people wouldn't want if they're not using it.

**Comment by jscheid** [comment]
Fair enough, the test setup is somewhat arcane. I'll take a look at this sometime soon.

You're right about `optionalDependencies`, but the code snippet behind the link above (without changes to package.json) does what we need I think. I.e. require in a try/catch, if it succeeds but doesn't have the right version print a warning and don't register the hook?

**Comment by jscheid** [comment]
Hmm scratch that, didn't have a close enough look. That example depends on the package exposing a `version` export which `html-webpack-plugin` doesn't.

**Comment by jscheid** [comment]
Can't see a good way to ensure that the right version is used, I guess people will have to pay attention to the README...

**Comment by jscheid** [comment]
@SPSpwetter perhaps something along these lines, https://gist.github.com/jscheid/dea42faec58f6e3083ad942a617563da

Can you think of other corner cases to test? Multiple entry points or such?

**Comment by jscheid** [comment]
@SPSpwetter see also my comment on the Gist.

**Comment by jscheid** [comment]
When it's all done, could you also add `Closes #3` to the commit message, then merging it will do just that.

Don't bother with it yet though, we should squash at the end anyway.

**Comment by SPSpwetter** [comment]
I'll add Closes #3 after I update the package.json to require ^2.21.0 of html-webpack-plugin.

**Comment by SPSpwetter** [comment]
I looked at stylesheet. That requires extract-text-webpack-plugin. Played with it a little bit but couldn't get it to work, so I backed out the changes until it matched what you gave me.

File: index.js [hunk 1]
```diff
@@ -75,6 +77,35 @@ SubresourceIntegrityPlugin.prototype.apply = function apply(compiler) {
...
+    }
+    if (!src) {
+      return;
+    }
+    var base = path.basename(src);
+    var asset = compilation.assets[base];
+    if (asset && asset.integrity) {
+      tag.attributes.integrity = asset.integrity;
+      tag.attributes.crossorigin = 'anonymous';
+    }
```
Comment (jscheid) [inline index.js]: I think this should be followed by something like

```
    } else {
      compilation.warnings.push(new Error(
        "webpack-subresource-integrity: cannot determine hash for asset '" +
          base + "', the resource will be unprotected."));
    }
```

Or perhaps even push to `compilation.errors` considering that all assets should come with a hash.

Comment (SPSpwetter) [inline index.js]: Hmmm... Ok. Original code didn't assume SRI was installed so it wouldn't complain if it didn't find asset.integrity, but if you're telling me that SRI will calculate a has for everything, then yeah, that seems like an error.

**Comment by jscheid** [comment]
https://gist.github.com/jscheid/dea42faec58f6e3083ad942a617563da#file-test-css-patch

**Comment by SPSpwetter** [comment]
That's pretty much what I was trying. But the test fails. 
Dumping the keys of result.compilation.assets gives me:` [ 'bundle.js', 'index.html' ]`

**Comment by jscheid** [comment]
Have you run `npm install` again? It silently fails when `style-loader` or `css-loader` are absent.

**Comment by SPSpwetter** [comment]
Doh! That was it.

**Comment by jscheid** [comment]
Thanks for working through this, I think this is good now. Only small improvement I can think of is to add `html-webpack-plugin` to the tags in `package.json`.

Once the branch is released upstream why don't you change the version in package.json, squash, rebase, and include `Closes #3` in the commit message. I'll then merge and release version 0.4.0. Feel free to improve the tests in the meantime if anything else comes to mind but I think we're in a pretty good place now!  👍 🎉 😃

**Comment by jantimon** [comment]
This might be easier if you put it all in the compilation callback:

you could also write (untested):
```
compiler.plugin('compilation', function compilationPlugin(compilation) {
  function processTag(tag) {
    // Get asset path - src from scripts and href from links
    var src =  tag.attributes && (tag.attributes.href || tag.attributes.src);
    if (!src) {
      return;
    }
    var asset = compilation.assets[path.basename(src)];
    if (!asset || !asset.integrity) {
      compilation.errors.push(new Error(
         "webpack-subresource-integrity: cannot determine hash for asset '" +
         src + "', the resource will be unprotected."));
       return;
     }
     // Add integrity check sums
     tag.attributes.integrity = asset.integrity;
     tag.attributes.crossorigin = 'anonymous';
  }
  compilation.plugin('html-webpack-plugin-alter-asset-tags', function(pluginArgs, callback) {
    pluginArgs.head.forEach(processTag);
    pluginArgs.body.forEach(processTag);
    callback();
});
}
```

**Comment by jantimon** [comment]
Released html-webpack-plugin 2.21.0

**Comment by SPSpwetter** [comment]
Ok, fixed merge conflict. Now that the new release of html-webpack-plugin is out we should be good to go.

**Comment by jscheid** [comment]
Thanks, would you mind squashing, with the issue reference in the final commit message?

**Comment by SPSpwetter** [comment]
Ok, reading tutorial on squashing.

**Comment by jscheid** [comment]
`git rebase -i <base-branch>` and mark all commits except for the first one as `squash`.

**Comment by SPSpwetter** [comment]
Ok, I think that's right. FYI, you can set the repository up to squash PRs. https://help.github.com/articles/about-pull-request-merge-squashing/

**Comment by jscheid** [comment]
Yes, I know about that feature, just thought it would be easy for you to do. But it didn't work, there are still 24 commits in the PR. Never mind though, I'll set up squashing and do it myself, no big deal.

BTW there's special syntax for closing issues via PR's, see here https://help.github.com/articles/closing-issues-via-commit-messages/ -- but I can change the commit message when I squash.

**Comment by jscheid** [comment]
:shipit: 👍

**Comment by jscheid** [comment]
Published as 0.4.0, thanks again!
